Modules/_abc.c

```diff
@@ -0,0 +1,894 @@
+   Note: this counter is private. Use `abc.get_cache_token()` for
+   external code. */
+static PyObject *abc_invalidation_counter;
+
+typedef struct {
+    PyObject_HEAD
+    Py_ssize_t iterating;
+    PyObject *data; /* The actual set of weak references. */
+    PyObject *pending; /* Pending removals collected during iteration. */
+    PyObject *in_weakreflist;
+} _guarded_set;
+
+static void
+gset_dealloc(_guarded_set *self)
+{
+    Py_DECREF(self->data);
+    Py_DECREF(self->pending);
+    if (self->in_weakreflist != NULL) {
+        PyObject_ClearWeakRefs((PyObject *) self);
+    }
+    Py_TYPE(self)->tp_free(self);
+}
+
+static PyObject *
+gset_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
+{
+    PyObject *data = PySet_New(NULL);
+    if (data == NULL) {
+        return NULL;
+    }
+    PyObject *pending = PyList_New(0);
+    if (pending == NULL) {
+        Py_DECREF(data);
+        return NULL;
+    }
+    _guarded_set *self = (_guarded_set *) type->tp_alloc(type, 0);
+    if (self == NULL) {
+        Py_DECREF(data);
+        Py_DECREF(pending);
+        return NULL;
+    }
+    self->iterating = 0;
+    self->data = data;
+    self->pending = pending;
+    self->in_weakreflist = NULL;
+    return (PyObject *) self;
+}
+
+PyDoc_STRVAR(guarded_set_doc,
+"Internal weak set guarded against deletion during iteration.\n\
+Used by ABC machinery.");
+
+static PyTypeObject _guarded_set_type = {
+    PyVarObject_HEAD_INIT(NULL, 0)
+    "_guarded_set",                     /*tp_name*/
+    sizeof(_guarded_set),               /*tp_size*/
+    .tp_dealloc = (destructor)gset_dealloc,
+    .tp_flags = Py_TPFLAGS_DEFAULT,
+    .tp_weaklistoffset = offsetof(_guarded_set, in_weakreflist),
+    .tp_alloc = PyType_GenericAlloc,
+    .tp_new = gset_new,
+};
+
+/* This object stores internal state for ABCs.
+   Note that we can use normal sets for caches,
+   since they are never iterated over. */
+typedef struct {
+    PyObject_HEAD
+    _guarded_set *_abc_registry;
+    PyObject *_abc_cache; /* Normal set of weak references. */
+    PyObject *_abc_negative_cache; /* Normal set of weak references. */
+    PyObject *_abc_negative_cache_version;
+} _abc_data;
+
+static void
+abc_data_dealloc(_abc_data *self)
+{
+    Py_DECREF(self->_abc_registry);
+    Py_DECREF(self->_abc_cache);
+    Py_DECREF(self->_abc_negative_cache);
+    Py_DECREF(self->_abc_negative_cache_version);
+    Py_TYPE(self)->tp_free(self);
+}
+
+static PyObject *
+abc_data_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
+{
+    PyObject *registry = NULL, *cache = NULL, *negative_cache = NULL;
+    registry = gset_new(&_guarded_set_type, NULL, NULL);
+    if (registry == NULL) {
+        return NULL;
+    }
+    cache = PySet_New(NULL);
+    if (cache == NULL) {
+        Py_DECREF(registry);
+        return NULL;
+    }
+    negative_cache = PySet_New(NULL);
+    if (negative_cache == NULL) {
+        Py_DECREF(registry);
+        Py_DECREF(cache);
+        return NULL;
```

Comment:
If you replace `Py_DECREF` with `Py_CLEAR` in `error:` target you can just use `goto error;` in all "if" branches in this function, but up to you.

Reply:
I would stick with `DECREF`s just for consistency. We can update/polish this later I think.